dali/imgcodec/decoders/nvjpeg2k.cc

```diff
@@ -0,0 +1,191 @@
+  output_image.pitch_in_bytes = pitch_in_bytes;
+  if (ctx->bpp == 8) {
+    ctx->pixel_type = DALI_UINT8;
+    output_image.pixel_type = NVJPEG2K_UINT8;
+  } else {  // ctx->bpp == 16
+    ctx->pixel_type = DALI_UINT16;
+    output_image.pixel_type = NVJPEG2K_UINT16;
+  }
+  output_image.num_components = ctx->shape[2];
+
+  auto ret = nvjpeg2kDecode(nvjpeg2k_handle_, *ctx->nvjpeg2k_decode_state,
+                            *ctx->nvjpeg2k_stream, &output_image, *ctx->cuda_stream);
+
+  if (ret == NVJPEG2K_STATUS_SUCCESS) {
+    return true;
+  } else if (ret == NVJPEG2K_STATUS_BAD_JPEG || ret == NVJPEG2K_STATUS_JPEG_NOT_SUPPORTED) {
+    return false;
+  } else {
+    CUDA_CALL_EX(ret, in->SourceInfo());
+    DALI_FAIL("Unreachable");  // silence a warning
+  }
+}
+
+bool NvJpeg2000DecoderInstance::ConvertData(void *in,
+                                            uint8_t *out,
+                                            DecodeParams opts,
+                                            Context *ctx) {
+  if (opts.format == DALI_GRAY) {
+    // Converting to Gray, dropping alpha channels if needed
+    assert(ctx->shape[2] >= 3);
+    TYPE_SWITCH(ctx->pixel_type, type2id, Input, (uint8_t, uint16_t), (
+      PlanarRGBToGray<uint8_t, Input>(
+        out, reinterpret_cast<Input*>(in), ctx->pixels_count,
+        ctx->pixel_type, *ctx->cuda_stream);
+    ), DALI_FAIL(make_string("Unsupported input type: ", ctx->pixel_type)));  // NOLINT
+  } else {
+    // Converting to interleaved, dropping alpha channels if needed
+    assert(ctx->shape[2] >= 3);
+    TYPE_SWITCH(ctx->pixel_type, type2id, Input, (uint8_t, uint16_t), (
+      PlanarToInterleaved<uint8_t, Input>(
+        out, reinterpret_cast<Input*>(in), ctx->pixels_count,
+        ctx->req_nchannels, opts.format, ctx->pixel_type, *ctx->cuda_stream);
+    ), DALI_FAIL(make_string("Unsupported input type: ", ctx->pixel_type)));  // NOLINT
+  }
+  return true;
+}
+
+DecodeResult NvJpeg2000DecoderInstance::DecodeImplTask(int thread_idx,
+                                                       cudaStream_t stream,
+                                                       SampleView<GPUBackend> out,
+                                                       ImageSource *in,
+                                                       DecodeParams opts,
+                                                       const ROI &roi) {
+  Context ctx = {};
+  ctx.nvjpeg2k_decode_state = &thread_resources_[thread_idx].nvjpeg2k_decode_state;
+  ctx.nvjpeg2k_stream = &thread_resources_[thread_idx].nvjpeg2k_stream;
+  ctx.decode_event = &thread_resources_[thread_idx].decode_event;
+  ctx.cuda_stream = &thread_resources_[thread_idx].cuda_stream;
+  DecodeResult result = {false, nullptr};
+
+  if (!ParseJpeg2000Info(in, opts, &ctx))
+    return result;
+
+  CUDA_CALL(cudaEventSynchronize(*ctx.decode_event));
+
+  if (!ctx.needs_processing) {
+    result.success = DecodeJpeg2000(in, out.mutable_data<uint8_t>(), opts, &ctx);
```

Comment:
in line 172 we have a try/catch. Shouldn't the try/catch also wrap this DecodeJpeg2000 call?